src/uchen/tboard/tboard_file.cc

```diff
@@ -1,25 +1,167 @@
+#include <string>
+
+#ifdef _WIN32
+#include <winsock2.h>
+#endif
 
 #include "absl/log/log.h"
 
+#include "crc32c/crc32c.h"
+#include "proto/event.pb.h"
+#include "proto/summary.pb.h"
 #include "src/uchen/tboard/tboard_file.h"
 
+namespace {
+
+// MaskDelta is a fixed arbitrary number that is used to mask the CRC32
+// checksum. expected from TensorBoard.
+constexpr uint32_t kMaskDelta = 0xa282ead8ul;
+
+}
+
 namespace uchen::tboard {
 
-absl::StatusOr<TBoardFile> TBoardFile::Open(std::string_view path) {
-  std::ofstream file(std::string(path), std::ios::binary);
+absl::StatusOr<TBoardFile> TBoardFile::Open(std::string_view logs_directory,
+                                            std::string_view log_name) {
+  if (logs_directory.empty()) {
+    return absl::InvalidArgumentError("Log file name must be provided");
+  }
+
+  std::string timestamp =
+      absl::FormatTime("%Y%m%d-%H%M%S", absl::Now(), absl::UTCTimeZone());
+  std::filesystem::path events_directory =
+      std::filesystem::path(logs_directory) / std::filesystem::path(timestamp);
+
+  if (!std::filesystem::exists(events_directory)) {
+    if (!std::filesystem::create_directories(events_directory)) {
+      return absl::InternalError("Failed to create logs directory");
+    }
+  }
+
+  std::string hostname = "localhost";
+  char hostname_buffer[HOST_NAME_MAX + 1];
+  if (gethostname(hostname_buffer, sizeof(hostname_buffer)) == 0) {
+    hostname = hostname_buffer;
+  }
+
+  std::string time_in_seconds = std::to_string(time(nullptr));
+
+  std::string filename = absl::StrFormat("%s.out.tfevents.%s.%s", log_name,
+                                         time_in_seconds, hostname);
+  std::filesystem::path file_path =
+      events_directory / std::filesystem::path(filename);
+  std::ofstream file(file_path, std::ios::binary);
+
   if (!file.is_open()) {
     return absl::InternalError("Failed to open file");
   }
   return TBoardFile(std::move(file));
 }
 
-void TBoardFile::RecordLoss(float loss) {
-  LOG(INFO) << "Recording loss: " << loss;
-  file_ << "Loss: " << loss << std::endl;
+void TBoardFile::AddScalar(const std::string& tag, double value,
+                           uint32_t step) {
+  LOG(INFO) << "Adding scalar: " << tag << ": " << value << ", Step: " << value;
+
+  tensorflow::Summary summary;
+
+  tensorflow::Summary::Value* summary_value = summary.add_value();
+  summary_value->set_tag(tag);
+  summary_value->set_simple_value(value);
+
+  AddEvent(summary, step);
+}
+
+void TBoardFile::AddImage(const std::string tag, const std::string& image,
+                          uint32_t step, uint32_t width, uint32_t height,
+                          uint32_t channel, const std::string& image_name,
+                          const std::string& image_description) {
+  auto encoded_image = EncodeImage(image);
+  if (!encoded_image.ok()) {
+    LOG(ERROR) << "Failed to Add image \"" << image_name
+               << "\": " << encoded_image.status();
+    return;
```

Comment:
If there's an error, it needs to be propagated. AddImage should return `absl::Status`